File: java/src/main/java/benchmark/model/v2/Telemetry.java
package benchmark.model.v2;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public final class Telemetry implements Serializable {
  private static final long serialVersionUID = 1L;

  public String source;
  public long ts;
  public List<String> tags = new ArrayList<>();
  public double[] values = new double[0];

  public Telemetry() {}

  public Telemetry(String source, long ts, List<String> tags, double[] values) {
    this.source = source;
    this.ts = ts;
    this.tags = tags != null ? tags : new ArrayList<>();
    this.values = values != null ? values : new double[0];
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Telemetry t)) return false;
    if (ts != t.ts || !Objects.equals(source, t.source) || !Objects.equals(tags, t.tags)) {
      return false;
    }
    if (values.length != t.values.length) return false;
    for (int i = 0; i < values.length; i++) {
      if (Math.abs(values[i] - t.values[i]) > 1e-9) return false;
    }
    return true;
  }

  @Override
  public int hashCode() {
    int result = Objects.hash(source, ts, tags);
    result = 31 * result + Arrays.hashCode(values);
    return result;
  }
}

File: kotlin/src/main/kotlin/benchmark/model/v2/Telemetry.kt
package benchmark.model.v2

import com.squareup.moshi.JsonClass
import kotlinx.serialization.Serializable
import kotlinx.serialization.protobuf.ProtoNumber
import java.io.Serializable as JavaSerializable
import kotlin.math.abs

@Serializable
@JsonClass(generateAdapter = true)
data class Telemetry(
    @ProtoNumber(1) @JvmField var source: String = "",
    @ProtoNumber(2) @JvmField var ts: Long = 0L,
    @ProtoNumber(3) @JvmField var tags: MutableList<String> = mutableListOf(),
    @ProtoNumber(4) @JvmField var values: MutableList<Double> = mutableListOf(),
) : JavaSerializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Telemetry) return false
        if (ts != other.ts || source != other.source || tags != other.tags) return false
        if (values.size != other.values.size) return false
        for (i in values.indices) {
            if (abs(values[i] - other.values[i]) > 1e-5) return false
        }
        return true
    }

    override fun hashCode(): Int {
        var result = source.hashCode()
        result = 31 * result + ts.hashCode()
        result = 31 * result + tags.hashCode()
        result = 31 * result + values.hashCode()
        return result
    }
}
